File: test/contract/openapi_contract_test.go
package contract

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/c360/semstreams/component"
	"github.com/c360/semstreams/componentregistry"
	"gopkg.in/yaml.v3"
)

// OpenAPISpec represents the OpenAPI 3.0 specification structure
type OpenAPISpec struct {
	OpenAPI    string                 `yaml:"openapi"`
	Info       OpenAPIInfo            `yaml:"info"`
	Paths      map[string]interface{} `yaml:"paths"`
	Components OpenAPIComponents      `yaml:"components"`
}

type OpenAPIInfo struct {
	Title       string `yaml:"title"`
	Description string `yaml:"description"`
	Version     string `yaml:"version"`
}

type OpenAPIComponents struct {
	Schemas map[string]interface{} `yaml:"schemas"`
}

// loadOpenAPISpec loads and parses the OpenAPI spec with clear error messages
func loadOpenAPISpec(t *testing.T) *OpenAPISpec {
	t.Helper()

	repoRoot, err := findRepoRoot()
	if err != nil {
		t.Fatalf("Failed to find repository root: %v", err)
	}

	openapiPath := filepath.Join(repoRoot, "specs", "openapi.v3.yaml")

	// Check file exists
	if _, err := os.Stat(openapiPath); err != nil {
		t.Fatalf(`
OpenAPI spec not found at: %s

Solutions:
  1. Run 'task schema:generate' to generate OpenAPI spec
  2. Verify specs/ directory exists
  3. Set SEMSTREAMS_ROOT if running from unusual location

For more info, see: docs/OPENAPI_INTEGRATION.md
`, openapiPath)
	}

	// Load and parse OpenAPI spec
	data, err := os.ReadFile(openapiPath)
	if err != nil {
		t.Fatalf("Failed to read OpenAPI spec: %v", err)
	}

	var spec OpenAPISpec
	if err := yaml.Unmarshal(data, &spec); err != nil {
		t.Fatalf(`
Failed to parse OpenAPI spec: %v

This indicates corrupted or invalid YAML in the OpenAPI spec.

Solutions:
  1. Regenerate: task schema:generate
  2. Do not manually edit specs/openapi.v3.yaml
  3. Report issue if regeneration doesn't fix it

For more info, see: docs/SCHEMA_GENERATION.md
`, err)
	}

	return &spec
}

// TestCommittedOpenAPISpecValid validates the committed OpenAPI spec structure
func TestCommittedOpenAPISpecValid(t *testing.T) {
	spec := loadOpenAPISpec(t)

	// Validate OpenAPI version
	if spec.OpenAPI != "3.0.3" {
		t.Errorf("Invalid OpenAPI version: expected 3.0.3, got %s", spec.OpenAPI)
	}

	// Validate info section
	if spec.Info.Title == "" {
		t.Error("OpenAPI spec missing title")
	}
	if spec.Info.Version == "" {
		t.Error("OpenAPI spec missing version")
	}

	// Validate paths exist
	if len(spec.Paths) == 0 {
		t.Error("OpenAPI spec has no paths defined")
	}

	// Validate components/schemas exist
	if len(spec.Components.Schemas) == 0 {
		t.Error("OpenAPI spec has no component schemas defined")
	}
}

// TestOpenAPISpecContainsAllComponents validates all registered components are in OpenAPI spec
func TestOpenAPISpecContainsAllComponents(t *testing.T) {
	repoRoot, err := findRepoRoot()
	if err != nil {
		t.Fatalf("Failed to find repository root: %v", err)
	}

	openapiPath := filepath.Join(repoRoot, "specs", "openapi.v3.yaml")

	// Load OpenAPI spec as raw structure
	data, err := os.ReadFile(openapiPath)
	if err != nil {
		t.Fatalf("Failed to read OpenAPI spec: %v", err)
	}

	var spec map[string]interface{}
	if err := yaml.Unmarshal(data, &spec); err != nil {
		t.Fatalf("Failed to parse OpenAPI spec: %v", err)
	}

	// Initialize component registry
	registry := component.NewRegistry()
	if err := componentregistry.Register(registry); err != nil {
		t.Fatalf("Failed to register components: %v", err)
	}

	factories := registry.ListFactories()

	// Extract schema references from ComponentType.schema.oneOf
	components, ok := spec["components"].(map[string]interface{})
	if !ok {
		t.Fatal("OpenAPI spec missing components section")
	}

	schemas, ok := components["schemas"].(map[string]interface{})
	if !ok {
		t.Fatal("OpenAPI spec missing schemas section")
	}

	componentType, ok := schemas["ComponentType"].(map[string]interface{})
	if !ok {
		t.Fatal("OpenAPI spec missing ComponentType schema")
	}

	properties, ok := componentType["properties"].(map[string]interface{})
	if !ok {
		t.Fatal("ComponentType missing properties field")
	}

	schemaField, ok := properties["schema"].(map[string]interface{})
	if !ok {
		t.Fatal("ComponentType.properties missing schema field")
	}

	oneOf, ok := schemaField["oneOf"].([]interface{})
	if !ok {
		t.Fatal("ComponentType.properties.schema missing oneOf array")
	}

	// Build set of referenced schema files
	referencedSchemas := make(map[string]bool)
	for _, refItem := range oneOf {
		refMap, ok := refItem.(map[string]interface{})
		if !ok {
			continue
		}
		refStr, ok := refMap["$ref"].(string)
		if !ok {
			continue
		}
		// Extract filename from path like "../schemas/udp.v1.json"
		filename := filepath.Base(refStr)
		// Remove .v1.json to get component name
		if len(filename) > 8 && filename[len(filename)-8:] == ".v1.json" {
			componentName := filename[:len(filename)-8]
			referencedSchemas[componentName] = true
		}
	}

	// Verify each registered component is referenced in OpenAPI spec
	for name := range factories {
		if !referencedSchemas[name] {
			t.Errorf("Component %s not referenced in OpenAPI spec ComponentType.schema.oneOf", name)
		}
	}

	// Verify no extra references
	for name := range referencedSchemas {
		if _, exists := factories[name]; !exists {
			t.Errorf("OpenAPI spec references schema %s but component is not registered", name)
		}
	}
}

// TestOpenAPISpecPaths validates required API paths exist in OpenAPI spec
func TestOpenAPISpecPaths(t *testing.T) {
	repoRoot, err := findRepoRoot()
	if err != nil {
		t.Fatalf("Failed to find repository root: %v", err)
	}

	openapiPath := filepath.Join(repoRoot, "specs", "openapi.v3.yaml")

	// Load OpenAPI spec
	data, err := os.ReadFile(openapiPath)
	if err != nil {
		t.Fatalf("Failed to read OpenAPI spec: %v", err)
	}

	var spec map[string]interface{}
	if err := yaml.Unmarshal(data, &spec); err != nil {
		t.Fatalf("Failed to parse OpenAPI spec: %v", err)
	}

	paths, ok := spec["paths"].(map[string]interface{})
	if !ok {
		t.Fatal("OpenAPI spec missing paths section")
	}

	// Verify required paths exist
	// Note: Paths are service-relative (without /components prefix)
	// as they come from the service OpenAPI registry
	requiredPaths := []string{
		"/types",
		"/types/{id}",
		"/status/{name}",
		"/flowgraph",
		"/validate",
	}

	for _, path := range requiredPaths {
		if _, exists := paths[path]; !exists {
			t.Errorf("OpenAPI spec missing required path: %s", path)
		}
	}
}

// TestOpenAPISchemaReferences validates schema references point to existing files
func TestOpenAPISchemaReferences(t *testing.T) {
	repoRoot, err := findRepoRoot()
	if err != nil {
		t.Fatalf("Failed to find repository root: %v", err)
	}

	openapiPath := filepath.Join(repoRoot, "specs", "openapi.v3.yaml")
	schemasDir := filepath.Join(repoRoot, "schemas")

	// Load OpenAPI spec as raw structure
	data, err := os.ReadFile(openapiPath)
	if err != nil {
		t.Fatalf("Failed to read OpenAPI spec: %v", err)
	}

	var spec map[string]interface{}
	if err := yaml.Unmarshal(data, &spec); err != nil {
		t.Fatalf("Failed to parse OpenAPI spec: %v", err)
	}

	// Extract schema references from ComponentType.schema.oneOf
	components, ok := spec["components"].(map[string]interface{})
	if !ok {
		t.Fatal("OpenAPI spec missing components section")
	}

	schemas, ok := components["schemas"].(map[string]interface{})
	if !ok {
		t.Fatal("OpenAPI spec missing schemas section")
	}

	componentType, ok := schemas["ComponentType"].(map[string]interface{})
	if !ok {
		t.Fatal("OpenAPI spec missing ComponentType schema")
	}

	properties, ok := componentType["properties"].(map[string]interface{})
	if !ok {
		t.Fatal("ComponentType missing properties field")
	}

	schemaField, ok := properties["schema"].(map[string]interface{})
	if !ok {
		t.Fatal("ComponentType.properties missing schema field")
	}

	oneOf, ok := schemaField["oneOf"].([]interface{})
	if !ok {
		t.Fatal("ComponentType.properties.schema missing oneOf array")
	}

	// Verify each referenced schema file exists
	for _, refItem := range oneOf {
		refMap, ok := refItem.(map[string]interface{})
		if !ok {
			continue
		}
		refStr, ok := refMap["$ref"].(string)
		if !ok {
			continue
		}

		// Build full path to referenced file
		// Reference is like "../schemas/udp.v1.json"
		schemaPath := filepath.Join(schemasDir, filepath.Base(refStr))

		if _, err := os.Stat(schemaPath); err != nil {
			t.Errorf("OpenAPI spec references non-existent schema file: %s (resolved to %s)",
				refStr, schemaPath)
		}
	}
}

File: ui/src/lib/contract/openapi.contract.test.ts
import { describe, it, expect } from "vitest";
import { readFileSync, existsSync } from "fs";
import { join } from "path";
import YAML from "yaml";

/**
 * Contract tests for OpenAPI specification
 * These tests validate that the UI can load and use the committed OpenAPI spec
 * without needing the backend to be running
 */

// OpenAPI spec path must be provided via environment variable
// No default path - this enforces explicit configuration
const OPENAPI_SPEC_PATH = process.env.OPENAPI_SPEC_PATH || "";

const SCHEMAS_DIR = process.env.SCHEMAS_DIR || "";

/**
 * OpenAPI specification structure
 */
interface OpenAPISpec {
  openapi: string;
  info: {
    title: unknown;
    version: unknown;
    [key: string]: unknown;
  };
  paths: Record<
    string,
    {
      get?: {
        responses: Record<string, unknown>;
        [key: string]: unknown;
      };
      [key: string]: unknown;
    }
  >;
  components?: {
    schemas?: Record<
      string,
      {
        type?: unknown;
        properties?: {
          [key: string]: {
            oneOf?: Array<{ $ref?: string; [key: string]: unknown }>;
            [key: string]: unknown;
          };
        };
        required?: string[];
        [key: string]: unknown;
      }
    >;
    [key: string]: unknown;
  };
}

/**
 * Helper to read OpenAPI spec with clear error messages
 * Returns null if spec not found (tests will be skipped)
 */
function loadOpenAPISpec(): OpenAPISpec | null {
  try {
    if (!existsSync(OPENAPI_SPEC_PATH)) {
      console.log(
        `
⚠️  Skipping contract tests: OpenAPI spec not found at ${OPENAPI_SPEC_PATH}

These tests are optional. They validate the UI against a backend's OpenAPI spec.

To enable these tests:
  1. Set OPENAPI_SPEC_PATH to your backend's spec:
     export OPENAPI_SPEC_PATH=/path/to/your-backend/specs/openapi.v3.yaml
     npm test

  2. In a monorepo setup:
     export OPENAPI_SPEC_PATH=/path/to/backend/specs/openapi.v3.yaml
     npm test
			`.trim(),
      );
      return null; // Skip tests gracefully
    }

    const yamlContent = readFileSync(OPENAPI_SPEC_PATH, "utf8");
    return YAML.parse(yamlContent) as OpenAPISpec;
  } catch (err) {
    if (
      err instanceof Error &&
      err.message.includes("OpenAPI spec not found")
    ) {
      throw err; // Re-throw our custom error
    }
    const errorMessage = err instanceof Error ? err.message : String(err);
    throw new Error(`Failed to parse OpenAPI spec: ${errorMessage}`);
  }
}

/**
 * Helper to load schema file with clear error messages
 */
function loadSchemaFile(filename: string): Record<string, unknown> {
  const schemaPath = join(SCHEMAS_DIR, filename);

  try {
    if (!existsSync(schemaPath)) {
      throw new Error(
        `
Schema file not found: ${schemaPath}

Expected schemas directory: ${SCHEMAS_DIR}

Solutions:
  1. Ensure semstreams/schemas/ directory exists
  2. Run: task schema:generate (in semstreams repo)
  3. Set environment variable:
     export SCHEMAS_DIR=/path/to/schemas

For more info, see your backend's documentation on schema generation
			`.trim(),
      );
    }

    const schemaContent = readFileSync(schemaPath, "utf8");
    return JSON.parse(schemaContent) as Record<string, unknown>;
  } catch (err) {
    if (err instanceof Error && err.message.includes("Schema file not found")) {
      throw err; // Re-throw our custom error
    }
    const errorMessage = err instanceof Error ? err.message : String(err);
    throw new Error(`Failed to parse schema ${filename}: ${errorMessage}`);
  }
}

describe.skipIf(!existsSync(OPENAPI_SPEC_PATH))(
  "OpenAPI Specification Contract Tests",
  () => {
    let openapiSpec: OpenAPISpec | null;

    it("should load and parse the committed OpenAPI spec", () => {
      expect(() => {
        openapiSpec = loadOpenAPISpec();
      }).not.toThrow();

      expect(openapiSpec).toBeDefined();
      expect(openapiSpec).toBeTypeOf("object");
    });

    it("should have valid OpenAPI 3.0 structure", () => {
      openapiSpec = loadOpenAPISpec();
      if (!openapiSpec) {
        throw new Error("openapiSpec is null");
      }

      expect(openapiSpec.openapi).toBe("3.0.3");
      expect(openapiSpec.info).toBeDefined();
      expect(openapiSpec.info.title).toBeTruthy();
      expect(openapiSpec.info.version).toBeTruthy();
    });

    it("should have paths section", () => {
      openapiSpec = loadOpenAPISpec();
      if (!openapiSpec) {
        throw new Error("openapiSpec is null");
      }

      expect(openapiSpec.paths).toBeDefined();
      expect(Object.keys(openapiSpec.paths).length).toBeGreaterThan(0);
    });

    it("should have components section with schemas", () => {
      openapiSpec = loadOpenAPISpec();

      expect(openapiSpec?.components).toBeDefined();
      expect(openapiSpec?.components?.schemas).toBeDefined();
      expect(
        Object.keys(openapiSpec?.components?.schemas ?? {}).length,
      ).toBeGreaterThan(0);
    });

    it("should have ComponentType schema with component references", () => {
      openapiSpec = loadOpenAPISpec();
      if (!openapiSpec?.components?.schemas) {
        throw new Error("components.schemas is not defined");
      }

      const componentType = openapiSpec.components.schemas.ComponentType;
      expect(componentType).toBeDefined();
      expect(componentType?.type).toBe("object");
      expect(componentType?.properties).toBeDefined();
      expect(componentType?.properties?.schema).toBeDefined();
      expect(componentType?.properties?.schema?.oneOf).toBeDefined();
      expect(Array.isArray(componentType?.properties?.schema?.oneOf)).toBe(
        true,
      );
      expect(componentType?.properties?.schema?.oneOf?.length).toBeGreaterThan(
        0,
      );
    });

    it("should have required API paths for component management", () => {
      openapiSpec = loadOpenAPISpec();
      if (!openapiSpec) {
        throw new Error("openapiSpec is null");
      }

      const requiredPaths = [
        "/components/types",
        "/components/types/{id}",
        "/components/status/{name}",
        "/components/flowgraph",
        "/components/validate",
      ];

      for (const path of requiredPaths) {
        expect(openapiSpec.paths[path]).toBeDefined();
      }
    });

    it("should have GET method for /components/types", () => {
      openapiSpec = loadOpenAPISpec();
      if (!openapiSpec) {
        throw new Error("openapiSpec is null");
      }

      const typesPath = openapiSpec.paths["/components/types"];
      expect(typesPath).toBeDefined();
      expect(typesPath?.get).toBeDefined();
      expect(typesPath?.get?.responses["200"]).toBeDefined();
    });

    it("should reference component schemas in oneOf", () => {
      openapiSpec = loadOpenAPISpec();
      if (!openapiSpec?.components?.schemas) {
        throw new Error("components.schemas is not defined");
      }

      const componentType = openapiSpec.components.schemas.ComponentType;
      const oneOfRefs = componentType?.properties?.schema?.oneOf;
      if (!oneOfRefs) {
        throw new Error("oneOf refs not found");
      }

      // Each oneOf item should have a $ref to a component schema
      for (const item of oneOfRefs) {
        expect(item?.$ref).toBeDefined();
        expect(item?.$ref).toMatch(/\.\.\/schemas\/.*\.v1\.json$/);
      }
    });

    it("should have metadata fields in ComponentType", () => {
      openapiSpec = loadOpenAPISpec();
      if (!openapiSpec?.components?.schemas) {
        throw new Error("components.schemas is not defined");
      }

      const componentType = openapiSpec.components.schemas.ComponentType;
      expect(componentType?.properties?.id).toBeDefined();
      expect(componentType?.properties?.name).toBeDefined();
      expect(componentType?.properties?.type).toBeDefined();
      expect(componentType?.properties?.description).toBeDefined();
      expect(componentType?.properties?.protocol).toBeDefined();
      expect(componentType?.properties?.domain).toBeDefined();
      expect(componentType?.properties?.version).toBeDefined();
    });

    it("should have required fields specified in ComponentType", () => {
      openapiSpec = loadOpenAPISpec();
      if (!openapiSpec?.components?.schemas) {
        throw new Error("components.schemas is not defined");
      }

      const componentType = openapiSpec.components.schemas.ComponentType;
      expect(componentType?.required).toBeDefined();
      expect(Array.isArray(componentType?.required)).toBe(true);
      expect(componentType?.required).toContain("id");
      expect(componentType?.required).toContain("name");
      expect(componentType?.required).toContain("type");
    });
  },
);

describe.skipIf(!existsSync(OPENAPI_SPEC_PATH))(
  "OpenAPI Schema References",
  () => {
    it("should have all schema references pointing to existing files", () => {
      const openapiSpec = loadOpenAPISpec();
      if (!openapiSpec?.components?.schemas) {
        throw new Error("components.schemas is not defined");
      }

      const componentType = openapiSpec.components.schemas.ComponentType;
      const oneOfRefs = componentType?.properties?.schema?.oneOf;
      if (!oneOfRefs) {
        throw new Error("oneOf refs not found");
      }

      for (const item of oneOfRefs) {
        const ref = item?.$ref;
        if (!ref) continue;
        // ref is like "../schemas/udp.v1.json" relative to OpenAPI spec location
        const schemaFilename = ref.split("/").pop();
        if (!schemaFilename) continue;

        // loadSchemaFile will throw clear error if file doesn't exist
        expect(() => {
          loadSchemaFile(schemaFilename);
        }).not.toThrow();
      }
    });

    it("should have all referenced schemas be valid JSON", () => {
      const openapiSpec = loadOpenAPISpec();
      if (!openapiSpec?.components?.schemas) {
        throw new Error("components.schemas is not defined");
      }

      const componentType = openapiSpec.components.schemas.ComponentType;
      const oneOfRefs = componentType?.properties?.schema?.oneOf;
      if (!oneOfRefs) {
        throw new Error("oneOf refs not found");
      }

      for (const item of oneOfRefs) {
        const ref = item?.$ref;
        if (!ref) continue;
        const schemaFilename = ref.split("/").pop();
        if (!schemaFilename) continue;

        // loadSchemaFile parses JSON and throws clear error if invalid
        expect(() => {
          const schema = loadSchemaFile(schemaFilename);
          expect(schema).toBeDefined();
        }).not.toThrow();
      }
    });
  },
);
